File: uikit/src/main/java/com/sendbird/uikit/widgets/MessageRecyclerView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sendbird.uikit.R;
import com.sendbird.uikit.databinding.SbViewMessageRecyclerViewBinding;
import com.sendbird.uikit.utils.SoftInputUtils;
import com.sendbird.uikit.utils.TextUtils;

public class MessageRecyclerView extends FrameLayout {
    private final SbViewMessageRecyclerViewBinding binding;

    public MessageRecyclerView(@NonNull Context context) {
        this(context, null);
    }

    public MessageRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MessageRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.MessageListView, defStyle, 0);
        try {
            this.binding = SbViewMessageRecyclerViewBinding.inflate(LayoutInflater.from(getContext()), this, true);
            int dividerColor = a.getColor(R.styleable.MessageListView_sb_pager_recycler_view_divide_line_color, context.getResources().getColor(android.R.color.transparent));
            float dividerHeight = a.getDimension(R.styleable.MessageListView_sb_pager_recycler_view_divide_line_height, 0);
            int recyclerViewBackground = a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_background, R.color.background_50);
            int tooltipBackground = a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_tooltip_background, R.drawable.selector_tooltip_background_light);
            int tooltipTextAppearance = a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_tooltip_textappearance, R.style.SendbirdCaption1Primary300);
            int typingIndicatorTextAppearance = a.getResourceId(R.styleable.MessageListView_sb_message_typing_indicator_textappearance, R.style.SendbirdCaption1OnLight02);
            int bannerBackground = a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_banner_background, R.drawable.sb_shape_channel_information_bg);
            int bannerTextAppearance = a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_banner_textappearance, R.style.SendbirdCaption2OnLight01);

            int scrollBottomBackground = a.getResourceId(R.styleable.MessageListView_sb_message_scroll_bottom_background, R.drawable.selector_scroll_bottom_light);
            int scrollBottomIcon = a.getResourceId(R.styleable.MessageListView_sb_message_scroll_bottom_icon, R.drawable.icon_chevron_down);
            ColorStateList scrollBottomTintColor = a.getColorStateList(R.styleable.MessageListView_sb_message_scroll_bottom_icon_tint);

            this.setBackgroundResource(android.R.color.transparent);
            binding.rvMessageList.setBackgroundResource(recyclerViewBackground);
            binding.rvMessageList.setOnTouchListener((v, event) -> {
                SoftInputUtils.hideSoftKeyboard(this);
                v.performClick();
                return false;
            });
            binding.rvMessageList.setUseDivider(false);
            binding.rvMessageList.setDividerColor(dividerColor);
            binding.rvMessageList.setDividerHeight(dividerHeight);

            binding.tvTooltipText.setBackgroundResource(tooltipBackground);
            binding.tvTooltipText.setTextAppearance(context, tooltipTextAppearance);
            binding.tvTypingIndicator.setTextAppearance(context, typingIndicatorTextAppearance);

            binding.ivScrollBottomIcon.setBackgroundResource(scrollBottomBackground);
            binding.ivScrollBottomIcon.setImageResource(scrollBottomIcon);
            binding.ivScrollBottomIcon.setImageTintList(scrollBottomTintColor);

            binding.tvBanner.setBackgroundResource(bannerBackground);
            binding.tvBanner.setTextAppearance(context, bannerTextAppearance);
        } finally {
            a.recycle();
        }
    }

    public void showTypingIndicator(@NonNull String text) {
        binding.tvTypingIndicator.setVisibility(View.VISIBLE);
        binding.tvTypingIndicator.setText(text);
    }

    public void hideTypingIndicator() {
        binding.tvTypingIndicator.setVisibility(View.GONE);
    }

    public void showNewMessageTooltip(@NonNull String text) {
        binding.vgTooltipBox.setVisibility(View.VISIBLE);
        binding.tvTooltipText.setText(text);
    }

    public void showScrollBottomButton() {
        binding.ivScrollBottomIcon.setVisibility(View.VISIBLE);
    }

    public void hideScrollBottomButton() {
        binding.ivScrollBottomIcon.setVisibility(View.GONE);
    }

    public void hideNewMessageTooltip() {
        binding.vgTooltipBox.setVisibility(View.GONE);
    }

    @NonNull
    public View getLayout() {
        return binding.getRoot();
    }

    @NonNull
    public PagerRecyclerView getRecyclerView() {
        return binding.rvMessageList;
    }

    @NonNull
    public View getTooltipView() {
        return binding.tvTooltipText;
    }

    @NonNull
    public View getScrollBottomView() {
        return binding.ivScrollBottomIcon;
    }

    @NonNull
    public View getTypingIndicator() {
        return binding.tvTypingIndicator;
    }

    @NonNull
    public View getBannerView() { return binding.tvBanner; }

    public void setBannerText(@Nullable String text) {
        binding.tvBanner.setVisibility(TextUtils.isEmpty(text) ? View.GONE : View.VISIBLE);
        binding.tvBanner.setText(text);
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/widgets/MessageRecyclerView.kt
package com.sendbird.uikit.internal.ui.widgets

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import com.sendbird.uikit.R
import com.sendbird.uikit.databinding.SbViewMessageRecyclerViewBinding
import com.sendbird.uikit.interfaces.OnConsumableClickListener
import com.sendbird.uikit.internal.extensions.setAppearance
import com.sendbird.uikit.utils.SoftInputUtils
import com.sendbird.uikit.utils.TextUtils

internal class MessageRecyclerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : FrameLayout(context, attrs, defStyle) {
    private val binding: SbViewMessageRecyclerViewBinding
    fun showTypingIndicator(text: String) {
        binding.tvTypingIndicator.visibility = VISIBLE
        binding.tvTypingIndicator.text = text
    }

    fun hideTypingIndicator() {
        binding.tvTypingIndicator.visibility = GONE
    }

    fun showNewMessageTooltip(text: String) {
        binding.vgTooltipBox.visibility = VISIBLE
        binding.tvTooltipText.text = text
    }

    fun showScrollFirstButton() {
        binding.ivScrollFirstIcon.visibility = VISIBLE
    }

    fun showUnreadCountTooltip(unreadCount: String) {
        binding.vgUnreadCountTooltip.visibility = VISIBLE
        binding.tvUnreadCountTooltipText.text = unreadCount
    }

    fun hideScrollFirstButton() {
        binding.ivScrollFirstIcon.visibility = GONE
    }

    fun hideNewMessageTooltip() {
        binding.vgTooltipBox.visibility = GONE
    }

    fun hideUnreadCountTooltip() {
        binding.vgUnreadCountTooltip.visibility = GONE
    }

    fun rotateScrollFirstButton(rotation: Float) {
        scrollFirstView.rotation = rotation
    }

    val layout: View
        get() = binding.root
    val recyclerView: PagerRecyclerView
        get() = binding.rvMessageList
    val tooltipView: View
        get() = binding.tvTooltipText
    val unreadCountTooltipIcon: View
        get() = binding.vgUnreadCountTooltipClose
    val scrollFirstView: View
        get() = binding.ivScrollFirstIcon
    val typingIndicator: View
        get() = binding.tvTypingIndicator
    val bannerView: View
        get() = binding.tvBanner
    var onScrollFirstButtonClickListener: OnConsumableClickListener? = null

    fun setBannerText(text: String?) {
        binding.tvBanner.visibility = if (TextUtils.isEmpty(text)) GONE else VISIBLE
        binding.tvBanner.text = text
    }

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.MessageListView, defStyle, 0)
        try {
            binding = SbViewMessageRecyclerViewBinding.inflate(LayoutInflater.from(getContext()), this, true)
            val dividerColor = a.getColor(
                R.styleable.MessageListView_sb_recycler_view_divide_line_color,
                ContextCompat.getColor(context, android.R.color.transparent)
            )
            val dividerHeight = a.getDimension(R.styleable.MessageListView_sb_recycler_view_divide_line_height, 0f)
            val recyclerViewBackground =
                a.getResourceId(R.styleable.MessageListView_sb_message_recyclerview_background, R.color.background_50)
            val tooltipBackground = a.getResourceId(
                R.styleable.MessageListView_sb_message_recyclerview_tooltip_background,
                R.drawable.selector_tooltip_background_light
            )
            val tooltipTextAppearance = a.getResourceId(
                R.styleable.MessageListView_sb_message_recyclerview_tooltip_textappearance,
                R.style.SendbirdCaption1Primary300
            )
            val unreadCountTooltipTextAppearance = a.getResourceId(
                R.styleable.MessageListView_sb_message_recyclerview_unread_count_tooltip_textappearance,
                R.style.SendbirdBody2OnLight02
            )
            val unreadCountTooltipIconTintColor =
                a.getColorStateList(R.styleable.MessageListView_sb_message_recyclerview_unread_count_tooltip_icon_tint)
            val typingIndicatorTextAppearance = a.getResourceId(
                R.styleable.MessageListView_sb_message_typing_indicator_textappearance,
                R.style.SendbirdCaption1OnLight02
            )
            val bannerBackground = a.getResourceId(
                R.styleable.MessageListView_sb_message_recyclerview_banner_background,
                R.drawable.sb_shape_channel_information_bg
            )
            val bannerTextAppearance = a.getResourceId(
                R.styleable.MessageListView_sb_message_recyclerview_banner_textappearance,
                R.style.SendbirdCaption2OnLight01
            )
            val scrollBottomBackground = a.getResourceId(
                R.styleable.MessageListView_sb_message_scroll_bottom_background,
                R.drawable.selector_scroll_bottom_light
            )
            val scrollBottomIcon =
                a.getResourceId(R.styleable.MessageListView_sb_message_scroll_bottom_icon, R.drawable.icon_chevron_down)
            val scrollBottomTintColor =
                a.getColorStateList(R.styleable.MessageListView_sb_message_scroll_bottom_icon_tint)
            setBackgroundResource(android.R.color.transparent)
            binding.rvMessageList.setBackgroundResource(recyclerViewBackground)
            binding.rvMessageList.setOnTouchListener { v: View, _: MotionEvent? ->
                SoftInputUtils.hideSoftKeyboard(this)
                v.performClick()
                false
            }
            binding.rvMessageList.setUseDivider(false)
            binding.rvMessageList.setDividerColor(dividerColor)
            binding.rvMessageList.setDividerHeight(dividerHeight)
            binding.tvTooltipText.setBackgroundResource(tooltipBackground)
            binding.tvTooltipText.setAppearance(context, tooltipTextAppearance)
            binding.vgUnreadCountTooltip.setBackgroundResource(tooltipBackground)
            binding.tvUnreadCountTooltipText.setAppearance(context, unreadCountTooltipTextAppearance)
            binding.ivUnreadCountTooltipClose.imageTintList = unreadCountTooltipIconTintColor
            binding.tvTypingIndicator.setAppearance(context, typingIndicatorTextAppearance)
            binding.ivScrollFirstIcon.setBackgroundResource(scrollBottomBackground)
            binding.ivScrollFirstIcon.setImageResource(scrollBottomIcon)
            binding.ivScrollFirstIcon.imageTintList = scrollBottomTintColor
            binding.ivScrollFirstIcon.setOnClickListener {
                if (onScrollFirstButtonClickListener?.onClick(it) == true) return@setOnClickListener
                recyclerView.stopScroll()
                recyclerView.scrollToPosition(0)
            }
            binding.tvBanner.setBackgroundResource(bannerBackground)
            binding.tvBanner.setAppearance(context, bannerTextAppearance)
        } finally {
            a.recycle()
        }
    }
}
